cpp/serve/grammar/grammar_state_matcher_base.h

```diff
@@ -0,0 +1,206 @@
+/*!
+ *  Copyright (c) 2023 by Contributors
+ * \file serve/grammar/grammar_state_matcher.cc
+ */
+#ifndef MLC_LLM_SERVE_GRAMMAR_GRAMMAR_STATE_MATCHER_BASE_H_
+#define MLC_LLM_SERVE_GRAMMAR_GRAMMAR_STATE_MATCHER_BASE_H_
+
+#include <vector>
+
+#include "../../tokenizers.h"
+#include "grammar.h"
+#include "grammar_state_matcher_state.h"
+
+namespace mlc {
+namespace llm {
+namespace serve {
+
+using namespace tvm::runtime;
+
+/*! \brief Check the codepoint is contained in the character class. */
+inline bool CharacterClassContains(const BNFGrammarNode::RuleExpr& rule_expr,
+                                   TCodepoint codepoint) {
+  DCHECK(rule_expr.type == BNFGrammarNode::RuleExprType::kCharacterClass ||
+         rule_expr.type == BNFGrammarNode::RuleExprType::kNegCharacterClass);
+  for (int i = 0; i < rule_expr.size(); i += 2) {
+    if (rule_expr.data[i] <= codepoint && codepoint <= rule_expr.data[i + 1]) {
+      return rule_expr.type == BNFGrammarNode::RuleExprType::kCharacterClass;
+    }
+  }
+  return rule_expr.type == BNFGrammarNode::RuleExprType::kNegCharacterClass;
+}
+
+/* \brief The concrete implementation of GrammarStateMatcherNode. */
+class GrammarStateMatcherBase {
+ protected:
+  using RuleExpr = BNFGrammarNode::RuleExpr;
+  using RuleExprType = BNFGrammarNode::RuleExprType;
+
+ public:
+  GrammarStateMatcherBase(const BNFGrammar& grammar, RulePosition init_rule_position = {})
+      : grammar_(grammar), tree_(grammar), stack_tops_history_(&tree_) {
+    InitStackState(init_rule_position);
+  }
+
+  bool AcceptCodepoint(TCodepoint codepoint, bool verbose = false) {
+    if (verbose) {
+      std::cout << "Stack before accepting: " << PrintStackState() << std::endl;
+    }
+    static std::vector<int32_t> new_stack_tops;
```

Comment:
avoid static members in function, instead, make it part of class member. Having static variables makes the code not threadsafe, aka we cannot create multiple GrammarStateMatcher in different threads and run them in parallel